File: eval-reviews/2026-09-12-chained/query-null/python/sol/r1/starter/binder.py
"""Resolve all names and types before optimization or row evaluation."""
from model import Plan, require
from parser import AGGREGATES


def walk(e):
    yield e
    for a in e.args: yield from walk(a)


def bind(q, database):
    aliases = [a for _, a in q.select]
    require(len(set(aliases)) == len(aliases), 'DUPLICATE_ALIAS')
    require(len(set(a for _, a in q.sources)) == len(q.sources), 'DUPLICATE_ALIAS')
    tables, columns = [], []
    for si, (name, alias) in enumerate(q.sources):
        require(name in database, 'UNKNOWN_TABLE')
        t = database[name]
        tables.append(t)
        columns.extend((alias, c['name'], c['type'], si) for c in t['columns'])
        if si: check(q.joins[si - 1][1], columns, False, False, True)
    for e in q.groups: check(e, columns, False)
    keys = [e.index for e in q.groups]
    require(len(set(keys)) == len(keys), 'INVALID_AGGREGATION')
    if q.where: check(q.where, columns, False, False, True)
    for e, _ in q.select: check(e, columns, True)
    if q.having: check(q.having, columns, True, False, True)
    roots = [e for e, _ in q.select] + ([q.having] if q.having else [])
    aggregate = bool(q.groups) or any(e.op in AGGREGATES for root in roots for e in walk(root))
    require(not q.having or aggregate, 'INVALID_AGGREGATION')
    if aggregate:
        for root in roots: grouped(root, keys)
    q.order = [(aliases.index(a) if a in aliases else -1, d, n) for a, d, n in q.order]
    require(all(i >= 0 for i, _, _ in q.order), 'UNKNOWN_COLUMN')
    return Plan(q, tables, [[] for _ in tables], aggregate)


def grouped(e, keys):
    if e.op in AGGREGATES: return
    require(e.op != 'col' or e.index in keys, 'INVALID_AGGREGATION')
    for a in e.args: grouped(a, keys)


def check(e, columns, allow, inside=False, predicate=False):
    if e.op == 'col':
        qualifier, name = e.value
        matches = [(i, c) for i, c in enumerate(columns) if c[1] == name and (not qualifier or c[0] == qualifier)]
        require(bool(matches), 'UNKNOWN_COLUMN')
        require(len(matches) == 1, 'AMBIGUOUS_COLUMN')
        e.index, c = matches[0]
        e.type, e.source = c[2], c[3]
        e.types = {e.type}
    elif e.op == 'lit':
        e.types = {'int', 'text', 'bool'} if e.type == 'null' else {e.type}
    else:
        agg = e.op in AGGREGATES
        require(not agg or (allow and not inside), 'INVALID_AGGREGATION')
        for a in e.args: check(a, columns, allow, inside or agg)
        possibilities = [a.types for a in e.args]
        if e.op == 'COUNT': e.types = {'int'}
        elif e.op in ('SUM', '+', '-', '*'):
            require(all('int' in types for types in possibilities), 'TYPE_ERROR')
            e.types = {'int'}
        elif e.op in ('MIN', 'MAX'):
            e.types = possibilities[0] & {'int', 'text'}
            require(bool(e.types), 'TYPE_ERROR')
        elif e.op in ('NOT', 'AND', 'OR'):
            require(all('bool' in types for types in possibilities), 'TYPE_ERROR')
            e.types = {'bool'}
        elif e.op in ('IS_NULL', 'IS_NOT_NULL'):
            e.types = {'bool'}
        elif e.op == 'COALESCE':
            e.types = set.intersection(*possibilities)
            require(bool(e.types), 'TYPE_ERROR')
        else:
            compatible = set.intersection(*possibilities)
            if e.op not in ('=', '<>'): compatible &= {'int', 'text'}
            require(bool(compatible), 'TYPE_ERROR')
            e.types = {'bool'}
        e.type = next(iter(e.types)) if len(e.types) == 1 else 'null'
    require(not predicate or 'bool' in e.types, 'TYPE_ERROR')

File: eval-reviews/2026-09-12-chained/query-null/typescript/sol/r1/starter/binder.ts
/** Binding and static validation precede all rewrites, even on empty input. */
import { aggregates, requireThat as need } from "./model.ts";
import type { Expr, Query, Plan, Table, ScalarType } from "./model.ts";
interface BoundColumn {
  qualifier: string;
  name: string;
  type: ScalarType;
  source: number;
}
const allTypes: ScalarType[] = ["int", "text", "bool"];
function possible(e: Expr): ScalarType[] {
  return e.possible ?? (e.type ? [e.type] : allTypes);
}
function setTypes(e: Expr, types: ScalarType[]): void {
  e.possible = types;
  e.type = types.length === 1 ? types[0] : undefined;
}
function common(args: Expr[], allowed = allTypes): ScalarType[] {
  return allowed.filter((t) => args.every((a) => possible(a).includes(t)));
}
export function* walk(e: Expr): Generator<Expr> {
  yield e;
  for (const a of e.args) yield* walk(a);
}
function check(
  e: Expr,
  columns: BoundColumn[],
  allow: boolean,
  inside = false,
  predicate = false,
): void {
  if (e.op === "col") {
    const [qualifier, name] = e.value as [string, string];
    const matches = columns
      .map((c, i) => ({ c, i }))
      .filter(
        ({ c }) => c.name === name && (!qualifier || c.qualifier === qualifier),
      );
    need(matches.length > 0, "UNKNOWN_COLUMN");
    need(matches.length === 1, "AMBIGUOUS_COLUMN");
    const { c, i } = matches[0];
    e.index = i;
    e.type = c.type;
    e.possible = [c.type];
    e.source = c.source;
  } else if (e.op === "lit") {
    e.possible = e.type ? [e.type] : allTypes;
  } else {
    const agg = aggregates.has(e.op);
    need(!agg || (allow && !inside), "INVALID_AGGREGATION");
    for (const a of e.args) check(a, columns, allow, inside || agg);
    if (e.op === "COUNT") setTypes(e, ["int"]);
    else if (["SUM", "+", "-", "*"].includes(e.op)) {
      need(e.args.every((a) => possible(a).includes("int")), "TYPE_ERROR");
      setTypes(e, ["int"]);
    } else if (["MIN", "MAX"].includes(e.op)) {
      const types = common(e.args, ["int", "text"]);
      need(types.length, "TYPE_ERROR");
      setTypes(e, types);
    } else if (e.op === "COALESCE") {
      const types = common(e.args);
      need(types.length, "TYPE_ERROR");
      setTypes(e, types);
    } else if (e.op === "IS NULL" || e.op === "IS NOT NULL") {
      setTypes(e, ["bool"]);
    } else if (["NOT", "AND", "OR"].includes(e.op)) {
      need(e.args.every((a) => possible(a).includes("bool")), "TYPE_ERROR");
      setTypes(e, ["bool"]);
    } else {
      const allowed: ScalarType[] = ["=", "<>"].includes(e.op)
        ? allTypes
        : ["int", "text"];
      need(common(e.args, allowed).length, "TYPE_ERROR");
      setTypes(e, ["bool"]);
    }
  }
  need(!predicate || possible(e).includes("bool"), "TYPE_ERROR");
}
function grouped(e: Expr, keys: Set<number>): void {
  if (aggregates.has(e.op)) return;
  need(e.op !== "col" || keys.has(e.index!), "INVALID_AGGREGATION");
  for (const a of e.args) grouped(a, keys);
}
export function bind(q: Query, database: Map<string, Table>): Plan {
  const aliases = q.select.map(([, a]) => a);
  need(new Set(aliases).size === aliases.length, "DUPLICATE_ALIAS");
  need(
    new Set(q.sources.map(([, a]) => a)).size === q.sources.length,
    "DUPLICATE_ALIAS",
  );
  const tables: Table[] = [],
    columns: BoundColumn[] = [];
  q.sources.forEach(([name, qualifier], source) => {
    const t = database.get(name);
    need(t, "UNKNOWN_TABLE");
    tables.push(t);
    columns.push(...t.columns.map((c) => ({ ...c, qualifier, source })));
    if (source) check(q.joins[source - 1].on, columns, false, false, true);
  });
  q.groups.forEach((e) => check(e, columns, false));
  const keys = new Set(q.groups.map((e) => e.index!));
  need(keys.size === q.groups.length, "INVALID_AGGREGATION");
  if (q.where) check(q.where, columns, false, false, true);
  q.select.forEach(([e]) => check(e, columns, true));
  if (q.having) check(q.having, columns, true, false, true);
  const roots = [...q.select.map(([e]) => e), ...(q.having ? [q.having] : [])];
  const nodes = roots.flatMap((e) => [...walk(e)]);
  const aggregate =
    q.groups.length > 0 || nodes.some((e) => aggregates.has(e.op));
  need(!q.having || aggregate, "INVALID_AGGREGATION");
  if (aggregate) roots.forEach((e) => grouped(e, keys));
  q.order = q.order.map(([a, d, n]) => {
    const i = aliases.indexOf(a as string);
    need(i >= 0, "UNKNOWN_COLUMN");
    return [i, d, n];
  });
  return { query: q, tables, filters: tables.map(() => []), aggregate };
}
